### PR TITLE
Use Pydantic to model API requests/responses

### PR DESCRIPTION
**Pull request by shreddd**
This PR is meant to make the stack Pydantic friendly, including responses that match the Pydantic Entity models, validation etc. 

Also includes test code to enable docker container based testing for the above.

The PR ended up being a lot broader in scope and covers:

- [ ] Adding Pydantic models to responses
- [ ] Updated tests and test framework 
- [ ] docker compose steps for running ingest and test
- [ ] separation of docker-compose uv environments for test
- [ ] Adding Pydantic validation to ingest
- [ ] Miscellaneous code and doc cleanup

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -32,16 +32,20 @@ jobs:
...
+      # - name: Run tests locally
+      #   run: uv run pytest tests
 
       # Note: This spins up containers running the default services.
       - name: Spin up Docker Compose stack in background
         run: docker compose up --detach
 
       # Note: This spins up the "test" container.
+      - name: Run ingest script
+        run: docker compose --profile=tools run ingest
```
Comment (eecavanna) [inline .github/workflows/ci.yml:44]: Just passing through (I saw the "PR" number increase on the GitHub UI). This comment has become inconsistent with the code below it.

```suggestion
      # Note: This spins up the "ingest" container.
      - name: Run ingest script
        run: docker compose --profile=tools run ingest  
```

**Comment by eecavanna** [comment]
Hi @shreddd, I merged in PR https://github.com/ber-data/bertron/pull/76 a couple minutes ago. That has creates some Git merge conflicts on this branch. In case you want me to resolve those on this branch, you can message me here or on Slack and I'll take a crack at it.

**Comment by shreddd** [comment]
@eecavanna - I think we should reconcile testing approaches - I'm putting stuff in the top level tests dir rather than under src

**Comment by eecavanna** [comment]
I'll resolve the merge conflicts this evening.

**Comment by eecavanna** [comment]
I'll come back to this tomorrow. Tests are failing. I think the reason is — at least in part — that I'm using FastAPI's `TestClient` class ([docs](https://fastapi.tiangolo.com/reference/testclient/)) instead of a real HTTP client like `httpx`, while the API server is in a different container than the container running the tests. I'll switch to a real HTTP client tomorrow. I plan to use `httpx` (side note: I noticed some code is using `requests`, but `requests` is not documented as a dependency in `pyproject.toml`).

File: src/server.py [hunk 2]
```diff
@@ -59,7 +60,7 @@ def get_version() -> VersionResponse:
     )
 
 
-@app.get("/bertron")
+@app.get("/bertron", response_model=EntitiesResponse)
 def get_all_entities():
```
[the remark this answers is not in the document]
Comment (eecavanna) [reply src/server.py]: Implemented in commit a6cab92e9340c031483c3ee180f666f2efaffe27. Resolving thread.

File: docker-compose.yml [hunk 3]
```diff
@@ -9,8 +9,8 @@ services:
     #       host environment.
     # Docs: https://docs.docker.com/compose/how-tos/environment-variables/variable-interpolation/#additional-information
     environment:
-      MONGO_HOST: ${MONGO_HOST:?}
-      MONGO_PORT: ${MONGO_PORT:?}
+      MONGO_HOST: mongo
+      MONGO_PORT: 27017
```
Comment (eecavanna) [inline docker-compose.yml:13]: Note: I intentionally made these hard-coded. More info is in: https://github.com/ber-data/bertron/issues/78

**Comment by eecavanna** [comment]
I finished resolving the merge conflicts and getting the tests to pass.

**Comment by eecavanna** [comment]
Suggestion for PR title: "Use Pydantic to model API requests/responses"

File: tests/test_api.py [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+        # Verify ber_data_source is valid
+        valid_sources = ["EMSL", "ESS-DIVE", "NMDC", "JGI"]
+        assert entity["ber_data_source"] in valid_sources
+
+
+# Integration test that combines multiple operations
+class TestBertronAPIIntegration:
+    """Integration tests that combine multiple API operations."""
+    
+    base_url = "http://app:8000"
```
[the remark this answers is not in the document]
Comment (eecavanna) [reply tests/test_api.py:324]: I agree with this suggestion. The test is now using FastAPI's `TestClient`, which communicates directly with the FastAPI app instead of submitting real HTTP requests to an HTTP server.

From the [FastAPI docs](https://fastapi.tiangolo.com/reference/testclient/):
> You can use the `TestClient` class to test FastAPI applications without creating an actual HTTP and socket connection, just communicating directly with the FastAPI code.

**Comment by shreddd** [comment]
@eecavanna - ready for your review - feel free to fix small changes directly if needed.

**Comment by eecavanna** [comment]
Thanks! I'll review it this afternoon.

File: src/bertron_client.py [hunk 5]
```diff
@@ -6,7 +6,7 @@
 Provides methods to query and retrieve entity data from the BER data sources.
 """
 
-import requests
+import requests  # FIXME: `requests` is not listed as a dependency in `pyproject.toml`
```
[the remark this answers is not in the document]
Comment (eecavanna) [reply src/bertron_client.py:9]: I agree with this, but I don't plan to take this beyond adding that `FIXME`, since this is a "generated file."

**Comment by eecavanna** [comment]
Update: I think the issue was that my local (host) environment didn't have a Python virtual environment at `./.venv` (in the container, `/app/.venv`) at the time. I was able to get the container to stay running, by creating that Python virtual environment locally. This is a sign of a problem to me because the container is sharing the Python virtual environment with the host. I will fix this in `docker-compose.yml` now.

**Comment by shreddd** [comment]
Hmm - I wanted specific targets to avoid having to remember or cut and paste long commands. I would like to leave those in there, unless we want to switch to a makefile model which could also capture specific targets.

**Review by eecavanna — APPROVED** [review]
Hi @shreddd, thanks for implementing this. I think having the API ensure all the data it's sending and receiving fits into some Pydantic class will result in people finding the API easier to use.

Before merging, I suggest looking at the following commit, which I made just now (while waiting for an unrelated, long-running Mongo aggregation pipeline to finish running in another window):

https://github.com/ber-data/bertron/pull/77/commits/8022e99083d1659ce60f9b7d38d607eaa45a9efa

In that commit, I removed the requirement of manually running the ingest script before running the tests. Now, a test fixture will run the ingest script. The way I implemented it here, the database is destroyed and recreated between each test, for a clean slate. This could be optimized (i.e. more surgical) down the road, in case people find the test duration to be too long. An advantage of this approach over doing a single ingest before starting the tests, is that tests can freely manipulate the database (to test corner cases) without affecting subsequent tests or being affected by previous tests.
